File: team-frontend/src/lib/gateway/reviews.ts
import "server-only";

import type { Review } from "@/generated/platform/engagement/v1/engagement_pb.js";
import { makeClients } from "./client.js";
import { getToken } from "./session.js";

function gateway() {
  return makeClients(getToken());
}

export interface ViewReview {
  id: string;
  listingId: string;
  userId: string;
  userName: string;
  orderId: string;
  rating: number;
  comment: string;
  createdAt: string;
  mediaUrls: string[];
  helpfulCount: number;
  verifiedPurchase: boolean;
}

export interface ViewRatingBreakdown {
  star1: number;
  star2: number;
  star3: number;
  star4: number;
  star5: number;
}

export interface ViewRatingSummary {
  listingId: string;
  averageRating: number;
  reviewCount: number;
  breakdown: ViewRatingBreakdown;
}

export interface ViewShopRatingSummary {
  sellerId: string;
  averageRating: number;
  reviewCount: number;
  breakdown: ViewRatingBreakdown;
}

function mapReview(r: Review): ViewReview {
  let createdAt = "";
  if (r.createdAt) {
    createdAt = new Date(Number(r.createdAt.seconds) * 1000).toLocaleDateString(
      "vi-VN",
    );
  }
  return {
    id: r.id,
    listingId: r.listingId,
    userId: r.userId,
    userName: r.userName || "Người mua",
    orderId: r.orderId,
    rating: r.rating,
    comment: r.comment,
    createdAt,
    mediaUrls: r.mediaUrls ?? [],
    helpfulCount: Number(r.helpfulCount),
    verifiedPurchase: r.verifiedPurchase,
  };
}

export async function createReview(
  listingId: string,
  rating: number,
  comment: string,
  orderId?: string,
  mediaUrls: string[] = [],
): Promise<ViewReview> {
  const res = await gateway().engagement.createReview({
    listingId,
    rating,
    comment,
    orderId: orderId ?? "",
    mediaUrls,
  });
  if (!res.review) throw new Error("create review failed");
  return mapReview(res.review);
}

/** Mark a review helpful; returns the new helpful count (team-engagement
 * dedupes per user, so calling twice is a no-op there). */
export async function markReviewHelpful(reviewId: string): Promise<number> {
  const res = await gateway().engagement.markReviewHelpful({ reviewId });
  return Number(res.helpfulCount);
}

export async function getShopRatingSummary(
  sellerId: string,
): Promise<ViewShopRatingSummary> {
  try {
    const res = await gateway().engagement.getShopRatingSummary({ sellerId });
    const b = res.breakdown;
    return {
      sellerId: res.sellerId || sellerId,
      averageRating: res.averageRating || 5.0,
      reviewCount: Number(res.reviewCount),
      breakdown: {
        star1: b?.star1 ?? 0,
        star2: b?.star2 ?? 0,
        star3: b?.star3 ?? 0,
        star4: b?.star4 ?? 0,
        star5: b?.star5 ?? 0,
      },
    };
  } catch {
    return {
      sellerId,
      averageRating: 5.0,
      reviewCount: 0,
      breakdown: { star1: 0, star2: 0, star3: 0, star4: 0, star5: 0 },
    };
  }
}

export async function listReviews(
  listingId: string,
  ratingFilter = 0,
): Promise<ViewReview[]> {
  try {
    const res = await gateway().engagement.listReviews({
      listingId,
      ratingFilter,
    });
    return res.reviews.map(mapReview);
  } catch {
    return [];
  }
}

export async function getListingRatingSummary(
  listingId: string,
): Promise<ViewRatingSummary> {
  try {
    const res = await gateway().engagement.getListingRatingSummary({
      listingId,
    });
    const b = res.breakdown;
    return {
      listingId: res.listingId,
      averageRating: res.averageRating || 5.0,
      reviewCount: Number(res.reviewCount),
      breakdown: {
        star1: b?.star1 ?? 0,
        star2: b?.star2 ?? 0,
        star3: b?.star3 ?? 0,
        star4: b?.star4 ?? 0,
        star5: b?.star5 ?? 0,
      },
    };
  } catch {
    return {
      listingId,
      averageRating: 5.0,
      reviewCount: 0,
      breakdown: { star1: 0, star2: 0, star3: 0, star4: 0, star5: 0 },
    };
  }
}

File: team-engagement/internal/handler/reviews.go
package handler

import (
	"context"
	"errors"

	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"

	engagementv1 "github.com/buidangphuc/team-engagement/generated/platform/engagement/v1"
	"github.com/buidangphuc/team-engagement/internal/interceptor"
	"github.com/buidangphuc/team-engagement/internal/repository"
)

// MarkReviewHelpful records the caller's "helpful" vote (idempotent per user).
func (h *EngagementHandler) MarkReviewHelpful(
	ctx context.Context, req *engagementv1.MarkReviewHelpfulRequest,
) (*engagementv1.MarkReviewHelpfulResponse, error) {
	if err := interceptor.RequireScopes(ctx, "engagement:write"); err != nil {
		return nil, err
	}
	if req.GetReviewId() == "" {
		return nil, status.Error(codes.InvalidArgument, "review_id is required")
	}
	if h.reviewSvc == nil {
		return nil, status.Error(codes.Unimplemented, "reviews service unavailable")
	}

	count, err := h.reviewSvc.MarkHelpful(ctx, req.GetReviewId(), userID(ctx))
	if err != nil {
		if errors.Is(err, repository.ErrReviewNotFound) {
			return nil, status.Error(codes.NotFound, "review not found")
		}
		return nil, status.Errorf(codes.Internal, "mark helpful: %v", err)
	}
	return &engagementv1.MarkReviewHelpfulResponse{HelpfulCount: count}, nil
}

// GetShopRatingSummary aggregates ratings across a seller's listings.
func (h *EngagementHandler) GetShopRatingSummary(
	ctx context.Context, req *engagementv1.GetShopRatingSummaryRequest,
) (*engagementv1.GetShopRatingSummaryResponse, error) {
	if req.GetSellerId() == "" {
		return nil, status.Error(codes.InvalidArgument, "seller_id is required")
	}
	if h.reviewSvc == nil {
		return &engagementv1.GetShopRatingSummaryResponse{SellerId: req.GetSellerId()}, nil
	}

	summary, err := h.reviewSvc.GetShopRatingSummary(ctx, req.GetSellerId())
	if err != nil {
		return nil, status.Errorf(codes.Internal, "get shop rating summary: %v", err)
	}

	return &engagementv1.GetShopRatingSummaryResponse{
		SellerId:      summary.SellerID,
		AverageRating: summary.AverageRating,
		ReviewCount:   summary.ReviewCount,
		Breakdown: &engagementv1.RatingBreakdown{
			Star_1: summary.Breakdown.Star1,
			Star_2: summary.Breakdown.Star2,
			Star_3: summary.Breakdown.Star3,
			Star_4: summary.Breakdown.Star4,
			Star_5: summary.Breakdown.Star5,
		},
	}, nil
}
